text/0615-autotracking-memoization.md

```diff
@@ -0,0 +1,385 @@
+
+person.firstName = 'Jennifer';
+
+console.log(person.fullName); // Jennifer Weber
+console.log(count); // 2;
+```
+
+## Motivation
+
+Autotracking is the fundamental reactivity model within Ember Octane, and has
+been highly successful so far in its usage and adoption. However, users today
+can only integrate with autotracking via the `@tracked` decorator, which allows
+them to _create_ tracked root state. There is no way to write code that responds
+to changes in that root state directly - the only way to do so is indirectly via
+Ember's templating layer.
+
+An example of where users might want to do this is the [`@cached`](https://github.com/emberjs/rfcs/pull/566)
+decorator for getters. This decorator only reruns its code when the tracked
+state it accessed during its previous computation changes. Currently, it would
+not be possible to build this decorator with public APIs.
+
+For a more involved example, we can take a look at [ember-concurrency](http://ember-concurrency.com/),
+which allows users to define async tasks. Ember Concurrency has an `observes()`
+API for tasks, which allows tasks to rerun when a property changes. This API is
+not documented or encouraged, and instead lifecycle hooks are recommended to
+rerun tasks. However, in Octane, lifecycle hooks on components are no longer
+available, removing that as an option. A more ergonomic, autotracked version of
+concurrency tasks could be created if users had a way to react to changes in
+autotracked state.
+
+Data layers like Ember Data could also benefit from this capability. These
+layers tend to have to keep state in sync between multiple levels of caching,
+which traditionally was done with computed properties and eventing systems. The
+ability to use autotracking to replace these systems, and to define their own
+reactive semantics, could help complex libraries and data layers out immensely.
+
+## Detailed design
+
+This RFC proposes four functions to be added to Ember's public API:
+
+```ts
+interface Cache<T = unknown> {}
+
+function createCache<T>(fn: () => T): Cache<T>;
+
+function getValue<T>(cache: Cache<T>): T
+
+function isCache(maybeCache: object): boolean;
+
+function isConst(cache: Cache): boolean;
+```
+
+These functions are exposed as exports of the `@glimmer/tracking/primitives/cache`
+module:
+
+```ts
+import {
+  createCache,
+  getValue,
+  isCache,
+  isConst,
+} from '@glimmer/tracking/primitives/cache';
+```
+
+### Usage
+
+`cache` receives a function, and returns a cache instance for that function.
+Users can call `getValue()` on the cache instance to run the function and get
+the value of its output. The cache will then return the same value whenever
+`getValue` is called again, until one of the tracked values that was _consumed_
+while it was running previously has been _dirtied_.
+
+```ts
+class State {
+  @tracked value;
+}
+
+let state = new State();
+let count = 0;
+
+let counter = createCache(() => {
+  // consume the state
+  state.value;
```

Comment:
What about consuming multiple states, both of which are tracked?  This seems unnecessary in the context you have described in this RFC, but just curious!

```js
createCache(() => {
  state.value;
  state.otherValue;
});
```

Reply:
Right, if both of those values are tracked, and either of them were updated, then the cache would be invalidated.